webapp/src/components/systemConsole/index.tsx

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React, {useCallback, useMemo, useState} from 'react';
+
+import './style.scss';
+import EnableSurvey from 'components/systemConsole/enableSurvey';
+import Expiry from 'components/systemConsole/expiry/expiry';
+import Questions from 'components/systemConsole/questions/questions';
+import SurveyDateTime from 'components/systemConsole/surveyDateTime/surveyDateTime';
+import TeamFilter from 'components/systemConsole/teamFilter/teamFilter';
+
+import type {CombinedConfig, CustomComponentProps, CustomConfigTypes} from 'types/mattermost-webapp';
+
+export type CustomSettingChildComponentProp = CustomComponentProps & {
+    setInitialSetting: (settingId: string, settings: CustomConfigTypes) => void;
+}
+
+function SystemConsoleSetting(props: CustomComponentProps) {
+    const {id, onChange, setSaveNeeded} = props;
+
+    // This holds the combined config of all sub-configs
+    const [config, setConfig] = useState<CombinedConfig>({} as CombinedConfig);
+
+    const onChangeWrapper = useCallback((settingId: string, settings: CustomConfigTypes) => {
+        const newConfig = {
+            ...config,
+            [settingId]: settings,
+        };
+
+        setConfig(newConfig);
+        onChange(id, newConfig);
+        setSaveNeeded();
+    }, [config, id, onChange, setSaveNeeded]);
+
+    // This is called by sub-config components on mount.
+    // Every sub-config component sets its config either from the saved values
+    // or the default values. Whatever they do, they are expected
+    // to call this with the settings they decide are to be set initially.
+    //
+    // This is needed so that on change in one sub-setting, we know all other sub-settings as well and thus
+    // can save the entire setting object on server side.
+    // Otherwise the displayed default values will never get saved unless a user changes them.
+    const setDefaults = useCallback((settingId: string, settings: CustomConfigTypes) => {
+        setConfig((existingConfig) => ({
+            ...existingConfig,
+            [settingId]: settings,
+        }));
+    }, []);
```

Comment:
In order to simplify all this logic we could have this component (or another file) store all the defaults, and just use that. So oversimplifying everything...

constants.tsx
```
const DEFAULT_CONFIG = {
    key1: 'value1'
}
```
index.tsx:
```
import DEFAULT_CONFIG from './constants';

function mergeDefaultConfig(stateConfig) {
    const mergedConfig = {};
    for (key in DEFAULT_CONFIG) {
        mergedConfig[key] = stateConfig[key] ?? DEFAULT_CONFIG[key];
    }
    return mergedConfig;
}
const SystemConsoleSetting = (props) => {
    const [config, setConfig] = useState(() => mergeDefaultConfig(props.config....));
    const onChangeWrapper = useCallback((...) => {
        setConfig((existingConfig) => ({
            ...existingConfig,
            [settingId]: settings,
        }));
     }, []);

    ...
}
```
settingX.tsx:
```
import DEFAULT_CONFIG from './constants`
const SettingX = (props) => {
    const [localValue, setLocalValue] = useState(props.config..... ?? DEFAULT_CONFIG.settingX);
    ...
}
```

Reply:
It won't work when the default value itself is dynamic. For example the survey date's default value is today + 30 days. Later when we add the option to select survey cadence, the default value of date and time will change based on that setting.

I think its more flexible if each component managed its own default config for now as the requirement and scope can change. We're still in the POC stage kinda.